src/mame/drivers/cps1.cpp

```diff
@@ -390,24 +406,50 @@ READ16_MEMBER(cps_state::qsound_rom_r)
 
 READ16_MEMBER(cps_state::qsound_sharedram1_r)
 {
-	return m_qsound_sharedram1[offset] | 0xff00;
+	u16 ret = 0xffff;
+	if (ACCESSING_BITS_0_7)
+	{
+		if (!machine().side_effects_disabled())
+			m_maincpu->spin_until_time(attotime::from_nsec(100)); // MB8464A-10L (100ns)
```

Comment:
`m_maincpu->adjust_icount(n);` is the correct way to add waitstates in handlers, where `n` is a negative number. 
Regardless there are three underlying problems which applying this would probably cause more problems than the ones that [supposedly](https://mametesters.org/view.php?id=7443) [will ](https://mametesters.org/view.php?id=408) [eventually](https://mametesters.org/view.php?id=1989) fix:
1. Do you have the **exact** count for each accessed portion over testing the actual PCBs (and assuming that CPS1 and 2 have the same amount of penalty).
2. imho adding these into handlers is bad coding and should be avoided. Given how MAME framework currently works you may trampoline the full memory map range via an `address_bank_device` and apply waitstate algo before that, the best way would be to apply them shorthanded in the map children declarations instead.
3. waitstate actually applies on xtal clock, I believe that adjust_icount still applies on actual CPU speed instead. So if you have to divide by an uneven number you're out of luck and you are supposed to do that at CPU level, which proved to be annoying for Super NES already and unsure about if it works 100% right, it would require unit testing the machine which is TBD.

PS: I don't think that waitstate doesn't occur anyway if you read an open bus portion (like the upper bytes of the linked example).

Reply:
DTACK pin is asynchronous afaik. Yes the CPU will internally be stalled a multiple of XTAL clock cycles. But if you do a magic icount number it would mean the timing is wrong when cpu is overclocked.

Reply:
For CPS1, i've looked into the h/w side of this before, it's a bit more complex, as well as the fixed flipflop delays, the "PRG1" PAL can add an extra wait (the /ONEWAIT line, pin 14) based on address:
from prg1: `one_wait = (as & bgack & /a22) + (as & bgack &  a23)`
There's also provision for the B-board to override the line via CN2 pin D8 (/DTACK2 line) but I don't *think* any B-board makes use of it, but it's possible a certain game could do it's own thing.